### PR TITLE
Virtio_transitional: Add test cases for vsock device

### PR DESCRIPTION
**Pull request by vivianQizhu**
Depends on https://github.com/avocado-framework/avocado-vt/pull/2310 and https://github.com/avocado-framework/avocado-vt/pull/2311

Signed-off-by: Qianqian Zhu <qizhu@redhat.com>

### CODE REVIEW

**Comment by vivianQizhu** [comment]
@kylazhang Please help review, thanks. 
(1/6) type_specific.io-github-autotest-libvirt.virtio_transitional_vsock.default.boot_test.virtio: PASS (147.68 s)
 (2/6) type_specific.io-github-autotest-libvirt.virtio_transitional_vsock.default.boot_test.virtio_non_transitional: PASS (140.66 s)
 (3/6) type_specific.io-github-autotest-libvirt.virtio_transitional_vsock.rhel6_guest.default.boot_test.virtio_transitional: PASS (47.05 s)
 (4/6) type_specific.io-github-autotest-libvirt.virtio_transitional_vsock.rhel6_guest.default.hotplug_test.virtio_transitional: PASS (45.72 s)
 (5/6) type_specific.io-github-autotest-libvirt.virtio_transitional_vsock.rhel6_guest.with_pcie_to_pci_bridge.boot_test.virtio_transitional: PASS (47.55 s)
 (6/6) type_specific.io-github-autotest-libvirt.virtio_transitional_vsock.rhel6_guest.with_pcie_to_pci_bridge.hotplug_test.virtio_transitional: PASS (48.14 s)

File: libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    guest_src_url = params.get("guest_src_url")
+    virtio_model = params['virtio_model']
+    boot_with_vsock = (params.get('boot_with_vsock', 'yes') == 'yes')
+    hotplug = (params.get('hotplug', 'no') == 'yes')
+    addr_pattern = params['addr_pattern']
+    device_pattern = params['device_pattern']
+
+    def check_plug_to(bus_type='pcie-to-pci-bridge'):
+        """
+        Check if the nic is plugged onto pcie-to-pci-bridge
```
Comment (kylazhang) [inline libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py]: one blank line after

File: libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                                lspci_output)
+        if not device_str:
+            test.fail('lspci failed, no device "%s"' % device_pattern)
+
+    if guest_src_url:
+
+        def _download():
+            download_cmd = ("wget %s -O %s" % (guest_src_url, target_path))
+            if process.system(download_cmd, shell=True):
+                test.error("Failed to download file")
```
Comment (kylazhang) [inline libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py]: I just find we have avocado func for download usage - 
from avocado.utils import download
download.get_file(url, dest_image)

File: libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                          'controller_model': 'pcie-to-pci-bridge'}
+            libvirt.create_controller_xml(
+                contr_dict, "add_controller", vm_name)
+    vsock_xml = libvirt.create_vsock_xml(virtio_model)
+    if boot_with_vsock:
+        libvirt.add_vm_device(vmxml, vsock_xml)
+    try:
+        if (params["os_variant"] == 'rhel6' or
+                'rhel6' in params.get("shortname")):
+            iface_params = {'type': 'bridge', 'model': 'virtio-transitional'}
```
Comment (kylazhang) [inline libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py]: If not fixed type, please remove the {'type': 'bridge'}

**Comment by vivianQizhu** [comment]
@kylazhang Updated,
1. Add some comments to explain steps.
2. Utilize `download.get_file()`.
3. Remove `{'type': 'bridge'}`
Please help review, thanks.

**Comment by vivianQizhu** [comment]
@kylazhang Updated, thanks.
```
 (1/6) type_specific.io-github-autotest-libvirt.virtio_transitional_vsock.default.boot_test.virtio: PASS (44.43 s)
 (2/6) type_specific.io-github-autotest-libvirt.virtio_transitional_vsock.default.boot_test.virtio_non_transitional: PASS (44.92 s)
 (3/6) type_specific.io-github-autotest-libvirt.virtio_transitional_vsock.rhel6_guest.default.boot_test.virtio_transitional: PASS (46.74 s)
 (4/6) type_specific.io-github-autotest-libvirt.virtio_transitional_vsock.rhel6_guest.default.hotplug_test.virtio_transitional: PASS (42.45 s)
 (5/6) type_specific.io-github-autotest-libvirt.virtio_transitional_vsock.rhel6_guest.with_pcie_to_pci_bridge.boot_test.virtio_transitional: PASS (47.08 s)
 (6/6) type_specific.io-github-autotest-libvirt.virtio_transitional_vsock.rhel6_guest.with_pcie_to_pci_bridge.hotplug_test.virtio_transitional: PASS (44.26 s)
```

File: libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    """
+    vm_name = params.get("main_vm", "avocado-vt-vm1")
+    vm = env.get_vm(params["main_vm"])
+    vmxml = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    backup_xml = vmxml.copy()
+    add_pcie_to_pci_bridge = params.get("add_pcie_to_pci_bridge")
+    guest_src_url = params.get("guest_src_url")
+    virtio_model = params['virtio_model']
+    boot_with_vsock = (params.get('boot_with_vsock', 'yes') == 'yes')
+    hotplug = (params.get('hotplug', 'no') == 'yes')
```
Comment (Yingshun) [inline libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py]: we'd better update to
    boot_with_vsock = 'yes' == params.get('boot_with_vsock', 'yes')

File: libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    addr_pattern = params['addr_pattern']
+    device_pattern = params['device_pattern']
+
+    def check_plug_to(bus_type='pcie-to-pci-bridge'):
```
Comment (Yingshun) [inline libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py]: Looks like it should be a common func. Can you add it to avocado-vt?

Comment (vivianQizhu) [reply libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py]: @Yingshun Which module do you suggested to add to?

File: libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py (hunk 4 above)
Comment (vivianQizhu) [reply libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py]: @Yingshun Thanks for review, but any difference here?

File: libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py (hunk 5 above)
Comment (Yingshun) [reply libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py]: I think vm_xml is fine, @kylazhang any suggestions?

File: libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py (hunk 4 above)
Comment (Yingshun) [reply libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py]: No difference in functionality but It will make your code more readable.

Comment (vivianQizhu) [reply libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py]: @Yingshun I would like to keep it if no strong opinion here, I am used to current pattern more. :) Thanks.

File: libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py (hunk 5 above)
Comment (kylazhang) [reply libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py]: vm_xml's func are xml operation(get, set, del) for different component, and for check, if really common, can add in utils_test/libvirt.py, or just leave it in tp-libvirt
BTW, if you want to add in vm_xml, you can add get_vsock which have 1 vsock = self.xmltreefile.find('devices').find('vsock')

Comment (vivianQizhu) [reply libvirt/tests/src/virtio_transitional/virtio_transitional_vsock.py]: @Yingshun @kylazhang Not tested but what do you think about this way: https://github.com/autotest/tp-libvirt/pull/2392/files#diff-6cad3f19417ae20189a90aefda2de6db

File: libvirt/tests/src/virtio_transitional/virtio_transitional_base.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    return False
+
+
+def get_free_pci_slot(vm_name):
+    """
+    Get a free slot for given bus
+
+    :param bus_index: the index of the bus to be searched, e.g. 0x08
+    :param vmxml: vm_xml.VMXML object
+    :return: The first free slot of the bus
```
Comment (kylazhang) [inline libvirt/tests/src/virtio_transitional/virtio_transitional_base.py]: The docstring need to update

File: libvirt/tests/src/virtio_transitional/virtio_transitional_base.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                address.get('bus') == pci_bridge_index):
+            used_slot.append(address.get('slot'))
+    for slot_index in range(1, 30):
+        slot = "%0#4x" % slot_index
+        if slot not in used_slot:
+            return slot
+    return None
+
+
+def get_pci_bridge_index(vm_name):
```
Comment (kylazhang) [inline libvirt/tests/src/virtio_transitional/virtio_transitional_base.py]: Missing docstring

**Comment by vivianQizhu** [comment]
@kylazhang Updated, thanks.

**Pull request merged**